# To Find Coffee shop around Marina Bay Sands Hotel in Singapore

# 1. Import necessary Libraries

In [1]:
import requests
import pandas as pd
import numpy as np
import random

!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim

from IPython.display import Image
from IPython.core.display import HTML

from pandas.io.json import json_normalize

!conda install -c conda-forge folium=0.5.0 --yes
import folium

print('Folium installed')
print('Libraries imported')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    geopy-1.22.0               |     pyh9f0ad1d_0          63 KB  conda-forge
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0           conda-forge
    geopy:          

# 2. Define Foursquare Credentials and Version 
    

In [2]:
CLIENT_ID = '5D3PZH5IQAHFI5IL0VVN5042E3REOTG0Y3GZV1VSWJWFXFMH'
CLIENT_SECRET ='OFOC4ZKOVM1WE2ODHB5TECDGWY0R4YJOKYMNTZD451WZZ0ST'
VERSION = '20180604'
LIMIT = 30
print('My credentails:')
print('CLIENT_ID: '+CLIENT_ID)
print('CLIENT_SECRET: '+CLIENT_SECRET)

My credentails:
CLIENT_ID: 5D3PZH5IQAHFI5IL0VVN5042E3REOTG0Y3GZV1VSWJWFXFMH
CLIENT_SECRET: OFOC4ZKOVM1WE2ODHB5TECDGWY0R4YJOKYMNTZD451WZZ0ST


In [3]:
address ='10 Bayfront Ave, 018986,Singapore'
geolocator = Nominatim(user_agent='foursquare_agent')
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(latitude,longitude)

1.2836965 103.8607226


# 3. Search for a specific venue category

I would like to know how many Coffee shops around Marina Bay Sands Hotel

In [4]:
search_query = 'coffee'
radius = 1000
print(search_query + '...OK!')

coffee...OK!


In [5]:
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/search?client_id=5D3PZH5IQAHFI5IL0VVN5042E3REOTG0Y3GZV1VSWJWFXFMH&client_secret=OFOC4ZKOVM1WE2ODHB5TECDGWY0R4YJOKYMNTZD451WZZ0ST&ll=1.2836965,103.8607226&v=20180604&query=coffee&radius=1000&limit=30'

In [6]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ebbff4c95feaf001bd915a9'},
 'response': {'venues': [{'id': '4bfb8d0265fbc9b62e88916c',
    'name': 'The Coffee Bean & Tea Leaf Beanstro',
    'location': {'address': '#B2-20/20A, The Shoppes at Marina Bay Sands',
     'crossStreet': '2 Bayfront Ave',
     'lat': 1.2837743874259746,
     'lng': 103.85874281828215,
     'labeledLatLngs': [{'label': 'display',
       'lat': 1.2837743874259746,
       'lng': 103.85874281828215}],
     'distance': 220,
     'postalCode': '018972',
     'cc': 'SG',
     'city': 'Singapore',
     'country': 'Singapore',
     'formattedAddress': ['#B2-20/20A, The Shoppes at Marina Bay Sands (2 Bayfront Ave)',
      '018972',
      'Singapore']},
    'categories': [{'id': '4bf58dd8d48988d16d941735',
      'name': 'Café',
      'pluralName': 'Cafés',
      'shortName': 'Café',
      'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/cafe_',
       'suffix': '.png'},
      'primary': True}],
    'referralId': 'v-

Get relevant part of JSON and transform it into a pandas dataframe

In [7]:
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = json_normalize(venues)
dataframe.head()

categories  hasPerk  \
0  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
1  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
2  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
3  [{'id': '4bf58dd8d48988d142941735', 'name': 'A...    False   
4  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   

                         id                             location.address  \
0  4bfb8d0265fbc9b62e88916c  #B2-20/20A, The Shoppes at Marina Bay Sands   
1  505719e6e4b0e7a978d9b1f5       L1-71, The Shoppes at Marina Bay Sands   
2  4e4718b3c65bd6ffbe9bc495                 #B2-10, Marina Bay Link Mall   
3  4ff91753e4b0a7f7416d13e0                Supertree Grove Casual Dining   
4  4b4aac42f964a520a58c26e3                 #01-47B/53/54, Millenia Walk   

  location.cc location.city location.country location.crossStreet  \
0          SG     Singapore        Singapore       2 Bayfront Ave   
1          SG     Singapore        Singapore        Bayfront Ave.   
2          SG     Singapore        Singapore       8A Marina Blvd   
3          SG     Singapore        Singapore   Gardens by the Bay   
4          SG     Singapore        Singapore      9 Raffles Blvd.   

   location.distance                          location.formattedAddress  \
0                220  [#B2-20/20A, The Shoppes at Marina Bay Sands (...   
1                343  [L1-71, The Shoppes at Marina Bay Sands (Bayfr...   
2                880  [#B2-10, Marina Bay Link Mall (8A Marina Blvd)...   
3                450  [Supertree Grove Casual Dining (Gardens by the...   
4                866  [#01-47B/53/54, Millenia Walk (9 Raffles Blvd....   

                             location.labeledLatLngs  location.lat  \
0  [{'label': 'display', 'lat': 1.283774387425974...      1.283774   
1  [{'label': 'display', 'lat': 1.282831311159308...      1.282831   
2  [{'label': 'display', 'lat': 1.280677618359934...      1.280678   
3  [{'label': 'display', 'lat': 1.281531486140762...      1.281531   
4  [{'label': 'display', 'lat': 1.291429320572493...      1.291429   

   location.lng location.neighborhood location.postalCode location.state  \
0    103.858743                   NaN              018972            NaN   
1    103.857757                   NaN                 NaN            NaN   
2    103.853413                   NaN              018984            NaN   
3    103.864138          Marina South              018953            NaN   
4    103.859846                   NaN              039596            NaN   

                                  name    referralId  
0  The Coffee Bean & Tea Leaf Beanstro  v-1589379221  
1                  Trung Nguyen Coffee  v-1589379221  
2           The Coffee Bean & Tea Leaf  v-1589379221  
3              Hill Street Coffee Shop  v-1589379221  
4                          Coffee Club  v-1589379221

In [8]:
print(dataframe.shape)

(30, 18)


In [9]:
dataframe

categories  hasPerk  \
0   [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
1   [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
2   [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
3   [{'id': '4bf58dd8d48988d142941735', 'name': 'A...    False   
4   [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
5   [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...    False   
6   [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
7   [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
8   [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
9   [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
10  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...    False   
11  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
12  [{'id': '5665c7b9498e7d8a4f2c0f06', 'name': 'C...    False   
13  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
14  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
15  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
16  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...    False   
17  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
18  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...    False   
19  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
20  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...    False   
21  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
22  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...    False   
23  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
24  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...    False   
25  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
26  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...    False   
27  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...    False   
28  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...    False   
29  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...    False   

                          id  \
0   4bfb8d0265fbc9b62e88916c   
1   505719e6e4b0e7a978d9b1f5   
2   4e4718b3c65bd6ffbe9bc495   
3   4ff91753e4b0a7f7416d13e0   
4   4b4aac42f964a520a58c26e3   
5   4b6bd9b3f964a520791a2ce3   
6   541e9c69498e2b7ff8adbf7f   
7   59891a71bfc6d0756da3d4a1   
8   4b4aa64bf964a5203d8c26e3   
9   4d2ea35da6df6dcbf864db7a   
10  5b73ca0758002c002ce21252   
11  5dba22e533dff4000893f40c   
12  5cc6a4b4598e64002c54c097   
13  538df004498eecce5c710350   
14  4b822ab5f964a52033cc30e3   
15  52e1ef9211d281c5ff4e0c64   
16  54211d56498e9ab326c84021   
17  5c6223bdf1936e002ce3c76c   
18  5c06546a16ef67002c49d530   
19  569f5c09498efee6a1f35b5e   
20  4dcb2ceb45dd3831227fce09   
21  58ab889be0adac4d3a4f5260   
22  57cfd574498e62e82c0d0544   
23  4b7ff06ff964a520af4530e3   
24  59d31fc675eee43b36f354df   
25  570cdedd498e24ad4473d768   
26  59a372af002f4c019a0d5c43   
27  5024ec40e4b0a8a24d3d69ed   
28  4ce8d20db99754816010ff44   
29  5dc3b13124273d0008c00151   

                                     location.address location.cc  \
0         #B2-20/20A, The Shoppes at Marina Bay Sands          SG   
1              L1-71, The Shoppes at Marina Bay Sands          SG   
2                        #B2-10, Marina Bay Link Mall          SG   
3                       Supertree Grove Casual Dining          SG   
4                        #01-47B/53/54, Millenia Walk          SG   
5                         #01-13, The Singapore Flyer          SG   
6                                       Marina Square          SG   
7                                   One Raffles Place          SG   
8                              #01-106, Millenia Walk          SG   
9                                    One Raffles Quay          SG   
10                                  #B2-51, The Heart          SG   
11                              #01-92, Millenia Walk          SG   
12                                                NaN          SG   
13                          #04-31, O

Define information of interest and filter dataframe

In [10]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name             categories  \
0           The Coffee Bean & Tea Leaf Beanstro                   Café   
1                           Trung Nguyen Coffee                   Café   
2                    The Coffee Bean & Tea Leaf                   Café   
3                       Hill Street Coffee Shop       Asian Restaurant   
4                                   Coffee Club                   Café   
5               Robert's Coffee (Robert Paulig)            Coffee Shop   
6                    The Coffee Bean & Tea Leaf                   Café   
7                  Jewel Coffee (Raffles Place)                   Café   
8                    The Coffee Bean & Tea Leaf                   Café   
9                                   Coffee Club                   Café   
10                                 Jewel Coffee            Coffee Shop   
11                             Huggs Coffee Bar                   Café   
12                              Mellower Coffee  Corporate Coffee Shop   
13                        Three Cups Coffee Co.                   Café   
14                         San Francisco Coffee                   Café   
15                       SweetSpot Coffee Kiosk                   Café   
16                   The Coffee Bean & Tea Leaf            Coffee Shop   
17                         The Coffee Academics                   Café   
18                         The Coffee Academics            Coffee Shop   
19                   The Coffee BEAN & Tea Leaf                   Café   
20                                  Coffeesmith            Coffee Shop   
21                          Coffee Club Express                   Café   
22                              Dimbulah Coffee            Coffee Shop   
23                                o'Coffee Club                   Café   
24                              Dal.Komm Coffee            Coffee Shop   
25                             Starbucks Coffee                   Café   
26                          Local Coffee People            Coffee Shop   
27                                 Costa Coffee            Coffee Shop   
28  The Tiffany's Gourmet Coffee And Great Food                   Café   
29                    Five Oars Coffee Roasters            Coffee Shop   

                                              address  cc       city  \
0         #B2-20/20A, The Shoppes at Marina Bay Sands  SG  Singapore   
1              L1-71, The Shoppes at Marina Bay Sands  SG  Singapore   
2                        #B2-10, Marina Bay Link Mall  SG  Singapore   
3                       Supertree Grove Casual Dining  SG  Singapore   
4                        #01-47B/53/54, Millenia Walk  SG  Singapore   
5                         #01-13, The Singapore Flyer  SG  Singapore   
6                                       Marina Square  SG  Singapore   
7                                   One Raffles Place  SG        NaN   
8                              #01-106, Millenia Walk  SG  Singapore   
9                                    One Raffles Quay  SG  Singapore   
10                                  #B2-51, The Heart  SG  Singapore   
11                              #01-92, Millenia Walk  SG  Singapore   
12                                                NaN  SG  Singapore   
13                          #04-31, One Raffles Place  SG  Singapore   
14                                      Millenia Walk  SG  Singapore   
15             Level 1 Sands Expo & Convention Centre  SG  Singapore   
16                          #01-02/04, Esplanade Mall  SG        NaN   
17                        Marina Bay Link Mall #01-01  SG     Marina   
18                             Ocean financial Center  SG  Singapore   
19                                          Esplanade  SG  Singapore   
20                             62 Collyer Quay #02-11  SG  Singapore   
21       5 Raffles Place, Raffles Place MRT, B1-25/26  SG  Singapore   
22                        MBFC Tower 1, 8 Marina Blvd  SG  Singapore   
23                  

Let's visualize the Coffee Shop that are nearby

In [11]:
dataframe_filtered.name

0             The Coffee Bean & Tea Leaf Beanstro
1                             Trung Nguyen Coffee
2                      The Coffee Bean & Tea Leaf
3                         Hill Street Coffee Shop
4                                     Coffee Club
5                 Robert's Coffee (Robert Paulig)
6                      The Coffee Bean & Tea Leaf
7                    Jewel Coffee (Raffles Place)
8                      The Coffee Bean & Tea Leaf
9                                     Coffee Club
10                                   Jewel Coffee
11                               Huggs Coffee Bar
12                                Mellower Coffee
13                          Three Cups Coffee Co.
14                           San Francisco Coffee
15                         SweetSpot Coffee Kiosk
16                     The Coffee Bean & Tea Leaf
17                           The Coffee Academics
18                           The Coffee Academics
19                     The Coffee BEAN & Tea Leaf


In [12]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=15) # generate map centred around the Marina Bay Sands Hotel

# add a red circle marker to represent the Marina Bay Sands Hotel
folium.features.CircleMarker(
    [latitude, longitude],
    radius=10,
    color='red',
    popup='Marina Bay Sands Hotel',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)

# add the Coffee Shop as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map

# 4. Explore a Given Venue

In [13]:
venue_id = '4bfb8d0265fbc9b62e88916c' # ID of The Coffee Bean & Tea Leaf Beanstro
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)
url

'https://api.foursquare.com/v2/venues/4bfb8d0265fbc9b62e88916c?client_id=5D3PZH5IQAHFI5IL0VVN5042E3REOTG0Y3GZV1VSWJWFXFMH&client_secret=OFOC4ZKOVM1WE2ODHB5TECDGWY0R4YJOKYMNTZD451WZZ0ST&v=20180604'

In [14]:
result = requests.get(url).json()
print(result['response']['venue'].keys())
result['response']['venue']

dict_keys(['id', 'name', 'contact', 'location', 'canonicalUrl', 'categories', 'verified', 'stats', 'url', 'price', 'likes', 'dislike', 'ok', 'rating', 'ratingColor', 'ratingSignals', 'allowMenuUrlEdit', 'beenHere', 'specials', 'photos', 'reasons', 'description', 'storeId', 'hereNow', 'createdAt', 'tips', 'shortUrl', 'timeZone', 'listed', 'popular', 'seasonalHours', 'pageUpdates', 'inbox', 'parent', 'hierarchy', 'attributes', 'bestPhoto', 'colors'])


{'id': '4bfb8d0265fbc9b62e88916c',
 'name': 'The Coffee Bean & Tea Leaf Beanstro',
 'contact': {'phone': '+6566887001', 'formattedPhone': '+65 6688 7001'},
 'location': {'address': '#B2-20/20A, The Shoppes at Marina Bay Sands',
  'crossStreet': '2 Bayfront Ave',
  'lat': 1.2837743874259746,
  'lng': 103.85874281828215,
  'labeledLatLngs': [{'label': 'display',
    'lat': 1.2837743874259746,
    'lng': 103.85874281828215}],
  'postalCode': '018972',
  'cc': 'SG',
  'city': 'Singapore',
  'country': 'Singapore',
  'formattedAddress': ['#B2-20/20A, The Shoppes at Marina Bay Sands (2 Bayfront Ave)',
   '018972',
   'Singapore']},
 'canonicalUrl': 'https://foursquare.com/v/the-coffee-bean--tea-leaf-beanstro/4bfb8d0265fbc9b62e88916c',
 'categories': [{'id': '4bf58dd8d48988d16d941735',
   'name': 'Café',
   'pluralName': 'Cafés',
   'shortName': 'Café',
   'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/cafe_',
    'suffix': '.png'},
   'primary': True},
  {'id': '4bf58dd8d489

B. Get the venue's overall rating

In [15]:
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

5.4


In [16]:
venue_id = '505719e6e4b0e7a978d9b1f5' # ID of Trung Nguyen Coffee
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

6.4


In [17]:
venue_id = '4e4718b3c65bd6ffbe9bc495' # ID of The Coffee Bean & Tea Leaf
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

6.3


In [18]:
venue_id = '4ff91753e4b0a7f7416d13e0' # ID of Hill Street Coffee Shop
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

5.3


In [19]:
venue_id = '4b4aac42f964a520a58c26e3' # ID of Coffee Club
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

6.0


In [20]:
venue_id = '4b6bd9b3f964a520791a2ce3' # ID of Robert's Coffee
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

This venue has not been rated yet.


In [21]:
venue_id = '52e1ef9211d281c5ff4e0c64' # ID of SweetSpot Coffee Kiosk
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

This venue has not been rated yet.


Result is  Trung Nguyen Coffee where is highest rate in top 5 near Marina Bay Sands Hotel.

In [22]:
venue_id = '505719e6e4b0e7a978d9b1f5' # ID of Trung Nguyen Coffee
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

6.4


Get the number of tips

In [23]:
result['response']['venue']['tips']['count']

23

Create URL and send GET request. Make sure to set limit to get all tips

In [24]:
## Trung Nguyen Coffee Tips
limit = 15 # set limit to be greater than or equal to the total number of tips
url = 'https://api.foursquare.com/v2/venues/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION, limit)

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ebc01adf7706a001ba09373'},
 'response': {'tips': {'count': 23,
   'items': [{'id': '548bc165498e493218b332f8',
     'createdAt': 1418445157,
     'text': 'Crave for great Vietnamese Coffee in Sgp? Look no further as they serve the best Viet Coffee around. The top brew is the Legendee, their version of Luwak coffee at SGP16 per pop. Tres expensive.',
     'type': 'user',
     'canonicalUrl': 'https://foursquare.com/item/548bc165498e493218b332f8',
     'lang': 'en',
     'likes': {'count': 0, 'groups': []},
     'logView': True,
     'agreeCount': 0,
     'disagreeCount': 0,
     'todo': {'count': 0},
     'user': {'id': '84062387',
      'firstName': 'Richard',
      'lastName': 'F',
      'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
       'suffix': '/84062387-KZYY55DOIZBNIN0S.jpg'}}}]}}}

In [25]:
tips = results['response']['tips']['items']

tip = results['response']['tips']['items'][0]
tip.keys()

dict_keys(['id', 'createdAt', 'text', 'type', 'canonicalUrl', 'lang', 'likes', 'logView', 'agreeCount', 'disagreeCount', 'todo', 'user'])

Format column width and display all tips

In [26]:
pd.set_option('display.max_colwidth', -1)

tips_df = json_normalize(tips) # json normalize tips

# columns to keep
filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName', 'user.gender', 'user.id']
tips_filtered = tips_df.loc[:, filtered_columns]

# display tips
tips_filtered

/opt/conda/envs/Python36/lib/python3.6/site-packages/pandas/core/indexing.py:1494: FutureWarning: 
Passing list-likes to .loc or [] with any missing label will raise
KeyError in the future, you can use .reindex() as an alternative.

See the documentation here:
https://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike
  return self._getitem_tuple(key)


text  \
0  Crave for great Vietnamese Coffee in Sgp? Look no further as they serve the best Viet Coffee around. The top brew is the Legendee, their version of Luwak coffee at SGP16 per pop. Tres expensive.   

   agreeCount  disagreeCount                        id user.firstName  \
0  0           0              548bc165498e493218b332f8  Richard         

  user.lastName  user.gender   user.id  
0  F            NaN           84062387

# 5. Search a Foursquare User

Define URL, send GET request and display features associated with user

In [27]:

user_id = '84062387' # user ID with most agree counts and complete profile
ACCESS_TOKEN = 'MCCTLEAP524XYRCGG2NG2L10EYWW02N12RBL3JR0ZN5LMCO2'

url = 'https://api.foursquare.com/v2/users/{}?client_id={}&client_secret={}&oauth_token={}&v={}'.format(user_id, CLIENT_ID, CLIENT_SECRET, ACCESS_TOKEN, VERSION) # define URL

# send GET request
results = requests.get(url).json()
user_data = results['response']['user']

# display features associated with user
user_data.keys()

dict_keys(['id', 'firstName', 'lastName', 'gender', 'canonicalUrl', 'photo', 'friends', 'tips', 'homeCity', 'bio', 'contact', 'photos', 'checkinPings', 'pings', 'type', 'mayorships', 'checkins', 'lists', 'blockedStatus', 'lenses'])

In [28]:
print('First Name: ' + user_data['firstName'])
print('Last Name: ' + user_data['lastName'])
print('Home City: ' + user_data['homeCity'])

First Name: Richard
Last Name: Foo
Home City: Kuala Lumpur, KUL


In [29]:
user_data['tips']

{'count': 272}

Wow! So it turns out that Richard is a very active Foursquare user, with more than 272 tips

Get User's tips

In [30]:
# define tips URL
url = 'https://api.foursquare.com/v2/users/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format(user_id, CLIENT_ID, CLIENT_SECRET, VERSION, limit)

# send GET request and get user's tips
results = requests.get(url).json()
tips = results['response']['tips']['items']

# format column width
pd.set_option('display.max_colwidth', -1)

tips_df = json_normalize(tips)

# filter columns
filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id']
tips_filtered = tips_df.loc[:, filtered_columns]

# display user's tips
tips_filtered

text  \
0   They serve the best Grilled Unagi in KL. Secret lies in imported live Eels from Japan. They also have Taiwanese eels, but the Japanese option shines! End your meal with Mochi or Yuzu Sorbet.            
1   Great cosy, "homely" place serving Portuguese influenced Chinese food from Macao. Tops are Min Chi, Macao Pork Bun, & "Belacan" Pork Belly. End with  Kaya Toasts, Sawdust Pudding. Ask for specials.     
2   Food in general was ok. Cincalok Omelette, Chicken Pongteh, Pie Tee and Ikan Goreng Power were good but Okra came out as old as a museum piece. Staff were blur, and service could have been better.      
3   Refined, delicate, ultra fresh and awesome sushi courses from Chef Shingo Takahashi. Nothing less than stellar. Approx 25,000JPY for 40 pieces menu. Absolute bomb!                                       
4   The no frills restaurant here is a gem. Dishes are yummy and authentically Penang Nyonya. Penang Char Beehoon, Perut Ikan, Lobak, Ju Hu Char, Curry Fish Head, Assam Prawns & Char Kuey Teow stood out.   
5   Ask for the US Tomahawk Wagyu Ribeye. Perfectly grilled. The beef fats and leftover meats are used for a yummy Beef Char Kuey Teow ever! The Leng Chai Curry Hotpot is worth a try too.                   
6   Clams/mussels with many varieties of preparation are the main stars. Lemon Garlic Butter, Ginger and Lemongrass, Basilik, and Tumeric Butter are standouts. You can try the Summer Salad too.             
7   The board says it all : "Nodi" denotes the art of connecting people with coffee. And they do just that. Great robust tasting coffee, served at the correct temperature. Check it out.                     
8   Upscale cucina by BIG. Prices are steep, but it's worth a try. Go for Capellini, Truffle Tagliatelle, Seafood Risotto, and the Braised Shortrib. For meat lovers, ask for the Ribeye on the bone too.     
9   If you're looking for yummy caramelized char siew, crispy aromatic siew yoke, and great Hakka minced meat noodles, look no further. This is easily one of the better options available.                   
10  Seriously yummy pies here. Try the Pulled Lamb Pie and the Nacho's Pie with Cheesy Sauce. Make sure you add on the full works ie peas, mash and gravy as options. Try the Elvis or Choc Mint as finale.   
11  Two Michelin Stars outlet that serves one of the best sushis in Tokyo. Friendly and jovial service, the chefs explains the type of fish used in each morsel. USD200 for 45 pieces of heaven. Joy!         
12  They have been around for a good 60 years, and they dish out excellent Nyonya Cuisine. Try the signature Fried Roast Pork with Garlic, Assam Pomfret, Mui Choy Pork, Sharks Fin Omelette & Inchi Kabin.   
13  They serve seriously yummy and excellent food here. Check out Bitter Gourd Omelette, Braised Fish with Yam, Steamed Chicken, Tofu with Brinjal, and Sweet Vinegar Pork Ribs. Highly recommended.          
14  Tuck into possibly the best Popiah in Penang. Order the ones with added crab meat. The stock served with the Popiah is just heavenly. Wash all down with coconut water or ambra juice!                    

    agreeCount  disagreeCount                        id  
0   0           0              59ef4d67febf316609a5d64a  
1   2           0              59c7d3a2b6b04b0eb550492e  
2   0           0              597c23c635f98372270c4a6d  
3   0           0              59519fd0f62e0919c943e830  
4   1           0              592db102356b49697c706021  
5   2           0              58eee91d4928141b82d9a494  
6   2           0              5831bc2675e1373164f32660  
7   3           0              57f560a5498e64de28cff8c5  
8   0           0              57de0b6d498e0ed6c4f958c9  
9   7           0              57d8e5ce498eaf6eb20e30e4  
10  5           0              577ef85e498e8229ad108e86  
11  0           0              574f0290498e20a69caad49b  
12  8           0              57287985498e8432eaae7251  
13  2           0              57236a7a498eaed61f2284ef  
14  3           

Let's get the venue for the tip with the greatest number of agree counts

In [31]:
tip_id = '57287985498e8432eaae7251' # tip id is the greatest number of agree counts

# define URL
url = 'http://api.foursquare.com/v2/tips/{}?client_id={}&client_secret={}&v={}'.format(tip_id, CLIENT_ID, CLIENT_SECRET, VERSION)

# send GET Request and examine results
result = requests.get(url).json()
print(result['response']['tip']['venue']['name'])
print(result['response']['tip']['venue']['location'])

Tek Sen Restaurant (德盛飯店)
{'address': '18 & 20 Lebuh Carnarvon', 'lat': 5.417387344626408, 'lng': 100.33598219195335, 'labeledLatLngs': [{'label': 'display', 'lat': 5.417387344626408, 'lng': 100.33598219195335}], 'postalCode': '10100', 'cc': 'MY', 'city': 'Georgetown', 'state': 'Pulau Pinang', 'country': 'Malaysia', 'formattedAddress': ['18 & 20 Lebuh Carnarvon', '10100 Georgetown, Pulau Pinang', 'Malaysia']}


Get User's friends

In [32]:
user_friends = json_normalize(user_data['friends']['groups'][0]['items'])
user_friends

Empty DataFrame
Columns: []
Index: []

Interesting. Despite being very active, it turns out that Richard does not have any friends on Foursquare. This might definitely change in the fututre.

# Retrive the User's Profile Image

In [33]:
user_data

{'id': '84062387',
 'firstName': 'Richard',
 'lastName': 'Foo',
 'gender': 'male',
 'canonicalUrl': 'https://foursquare.com/user/84062387',
 'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
  'suffix': '/84062387-KZYY55DOIZBNIN0S.jpg'},
 'friends': {'count': 16,
  'groups': [{'type': 'friends',
    'name': 'Mutual friends',
    'count': 0,
    'items': []},
   {'type': 'others',
    'name': 'Other friends',
    'count': 16,
    'items': [{'id': '362240',
      'firstName': 'Becky',
      'lastName': 'Leng',
      'gender': 'female',
      'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
       'suffix': '/362240_pzKtltuD_jjS9_bZWb8y5e_g3DkG1HSIYmoO0Hgri0aSDhdk6unOzz4H3s1_ahd_IgykvIC95.jpg'},
      'tips': {'count': 8},
      'lists': {'groups': [{'type': 'created', 'count': 4, 'items': []}]},
      'homeCity': 'Kuala Lumpur, Malaysia',
      'bio': '',
      'contact': {'facebook': '592362588'}},
     {'id': '5364155',
      'firstName': 'Debbie',
      'lastName': 'Tan

In [34]:
# 1. grab prefix of photo
# 2. grab suffix of photo
# 3. concatenate them using the image size  
Image(url='https://fastly.4sqi.net/img/user/300x300/84062387-KZYY55DOIZBNIN0S.jpg')

# 6. Explore a location

In [47]:
latitude = 1.2828313111593086
longitude = 103.85775715419446
#'lat': 1.2828313111593086,
#'lng': 103.85775715419446,

Define URL

In [48]:
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=5D3PZH5IQAHFI5IL0VVN5042E3REOTG0Y3GZV1VSWJWFXFMH&client_secret=OFOC4ZKOVM1WE2ODHB5TECDGWY0R4YJOKYMNTZD451WZZ0ST&ll=1.2828313111593086,103.85775715419446&v=20180604&radius=1000&limit=30'

Send GET request and examine results

In [49]:
import requests

In [51]:
results = requests.get(url).json()
'There are {} around Trung Nguyen Coffee.'.format(len(results['response']['groups'][0]['items']))

'There are 30 around Trung Nguyen Coffee.'

In [52]:
items = results['response']['groups'][0]['items']
items[0]

{'reasons': {'count': 0,
  'items': [{'summary': 'This spot is popular',
    'type': 'general',
    'reasonName': 'globalInteractionReason'}]},
 'venue': {'id': '4dd8a3f9d1647fcf3e699bad',
  'name': 'Marina Bay Sands',
  'location': {'address': '10 Bayfront Avenue',
   'lat': 1.2831902557086854,
   'lng': 103.85912418365479,
   'labeledLatLngs': [{'label': 'display',
     'lat': 1.2831902557086854,
     'lng': 103.85912418365479}],
   'distance': 157,
   'postalCode': '018956',
   'cc': 'SG',
   'city': 'Singapore',
   'country': 'Singapore',
   'formattedAddress': ['10 Bayfront Avenue', '018956', 'Singapore']},
  'categories': [{'id': '4bf58dd8d48988d12f951735',
    'name': 'Resort',
    'pluralName': 'Resorts',
    'shortName': 'Resort',
    'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/travel/resort_',
     'suffix': '.png'},
    'primary': True}],
  'photos': {'count': 0, 'groups': []}},
 'referralId': 'e-0-4dd8a3f9d1647fcf3e699bad-0'}

Process JSON and convert it to a clean dataframe

In [53]:
dataframe = json_normalize(items) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# filter the category for each row
dataframe_filtered['venue.categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean columns
dataframe_filtered.columns = [col.split('.')[-1] for col in dataframe_filtered.columns]

dataframe_filtered.head(10)

name     categories             address  cc  \
0  Marina Bay Sands                 Resort         10 Bayfront Avenue  SG   
1  Marina Bay Sands Boardwalk       Waterfront     10 Bayfront Avenue  SG   
2  New Creation Church (MBS)        Church         Marina Bay Sands    SG   
3  The Shoppes at Marina Bay Sands  Shopping Mall  2 Bayfront Ave.     SG   
4  Marina Bay                       Bay            Marina Bay          SG   
5  Louis Vuitton Island Maison      Boutique       Marina Bay Sands    SG   
6  Marina Bay Sands Hotel           Hotel          10 Bayfront Ave.    SG   
7  Marina Bay Waterfront Promenade  Waterfront     Bayfront Ave        SG   
8  Art Stage Singapore              Art Gallery    Marina Bay Sands    SG   
9  CUT by Wolfgang Puck             Steakhouse     Marina Bay Sands    SG   

        city    country      crossStreet  distance  \
0  Singapore  Singapore  NaN              157        
1  Singapore  Singapore  NaN              158        
2  Singapore  Singapore  Bayfront Ave.    128        
3  Singapore  Singapore  NaN              207        
4  Singapore  Singapore  NaN              268        
5  Singapore  Singapore  2 Bayfront Ave.  268        
6  Singapore  Singapore  NaN              290        
7  Singapore  Singapore  NaN              309        
8  Singapore  Singapore  NaN              145        
9  Singapore  Singapore  10 Bayfront Ave  337        

                                          formattedAddress  \
0  [10 Bayfront Avenue, 018956, Singapore]                   
1  [10 Bayfront Avenue, 018956, Singapore]                   
2  [Marina Bay Sands (Bayfront Ave.), Singapore]             
3  [2 Bayfront Ave., 018972, Singapore]                      
4  [Marina Bay, Singapore]                                   
5  [Marina Bay Sands (2 Bayfront Ave.), 018972, Singapore]   
6  [10 Bayfront Ave., 018956, Singapore]                     
7  [Bayfront Ave, Singapore]                                 
8  [Marina Bay Sands, Singapore]                             
9  [Marina Bay Sands (10 Bayfront Ave), 018972, Singapore]   

                                                                 labeledLatLngs  \
0  [{'label': 'display', 'lat': 1.2831902557086854, 'lng': 103.85912418365479}]   
1  NaN                                                                            
2  [{'label': 'display', 'lat': 1.2824229508693408, 'lng': 103.85884036741291}]   
3  NaN                                                                            
4  [{'label': 'display', 'lat': 1.284992247519025, 'lng': 103.8566780090332}]     
5  [{'label': 'display', 'lat': 1.2851123562237643, 'lng': 103.85853752419233}]   
6  NaN                                                                            
7  [{'label': 'display', 'lat': 1.2811245137355043, 'lng': 103.85555642836434}]   
8  [{'label': 'display', 'lat': 1.283268182205801, 'lng': 103.8589875803299}]     
9  [{'label': 'display', 'lat': 1.2853498841123125, 'lng': 103.85944005306924}]   

        lat         lng neighborhood postalCode state  \
0  1.283190  103.859124  NaN          018956     NaN    
1  1.284048  103.858491  NaN          018956     NaN    
2  1.282423  103.858840  NaN          NaN        NaN    
3  1.283969  103.859237  NaN          018972     NaN    
4  1.284992  103.856678  NaN          NaN        NaN    
5  1.285112  103.858538  NaN          018972     NaN    
6  1.282932  103.860369  NaN          018956     NaN    
7  1.281125  103.855556  NaN          NaN        NaN    
8  1.283268  103.858988  NaN          NaN        NaN    
9  1.285350  103.859440  NaN          018972     NaN    

                         id  
0  4dd8a3f9d1647fcf3e699bad  
1  4ce55576ca41236ac9b37b4b  
2  4f93be18e4b02d4703686e14  
3  4bd6b4925631c9b60239a630  
4  527faef5498ecf5c6e90924f  
5  4e6ca596b993061ea8e39772  
6  4bf7c0404a67c928061924cf  
7  4c35c5693ffc9521753b91f5  
8  4d2d5e4388e23704c52339c7  
9  4d41cbdb9fa82c0f9a8ab08a

Let's visualize these items on the map around our location

In [54]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=25) # generate map centred around Trung Nguyen Coffee


# add Trung Nguyen Coffee as a red circle mark
folium.features.CircleMarker(
    [latitude, longitude],
    radius=10,
    popup='Trung Nguyen Coffee',
    fill=True,
    color='red',
    fill_color='red',
    fill_opacity=0.6
    ).add_to(venues_map)


# add popular spots to the map as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        fill=True,
        color='blue',
        fill_color='blue',
        fill_opacity=0.6
        ).add_to(venues_map)

# display map
venues_map

# 7. Explore Trending Venues

In [55]:
# define URL
url = 'https://api.foursquare.com/v2/venues/trending?client_id={}&client_secret={}&ll={},{}&v={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION)

# send GET request and get trending venues
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ebc04211a4b0a0028084f5f'},
 'response': {'venues': []}}

Check if any venues are trending at this time

In [56]:
if len(results['response']['venues']) == 0:
    trending_venues_df = 'No trending venues are available at the moment!'
    
else:
    trending_venues = results['response']['venues']
    trending_venues_df = json_normalize(trending_venues)

    # filter columns
    columns_filtered = ['name', 'categories'] + ['location.distance', 'location.city', 'location.postalCode', 'location.state', 'location.country', 'location.lat', 'location.lng']
    trending_venues_df = trending_venues_df.loc[:, columns_filtered]

    # filter the category for each row
    trending_venues_df['categories'] = trending_venues_df.apply(get_category_type, axis=1)

In [57]:
# display trending venues
trending_venues_df

'No trending venues are available at the moment!'

# Visualize trending venues

In [58]:
if len(results['response']['venues']) == 0:
    venues_map = 'Cannot generate visual as no trending venues are available at the moment!'

else:
    venues_map = folium.Map(location=[latitude, longitude], zoom_start=15) # generate map centred around Ecco


    # add Ecco as a red circle mark
    folium.features.CircleMarker(
        [latitude, longitude],
        radius=10,
        popup='Ecco',
        fill=True,
        color='red',
        fill_color='red',
        fill_opacity=0.6
    ).add_to(venues_map)


    # add the trending venues as blue circle markers
    for lat, lng, label in zip(trending_venues_df['location.lat'], trending_venues_df['location.lng'], trending_venues_df['name']):
        folium.features.CircleMarker(
            [lat, lng],
            radius=5,
            poup=label,
            fill=True,
            color='blue',
            fill_color='blue',
            fill_opacity=0.6
        ).add_to(venues_map)
# display map
venues_map

'Cannot generate visual as no trending venues are available at the moment!'

In [59]:
print('End Programe')

End Programe
